File: app/src/main/java/ch/bailu/aat/services/directory/FilesInDirectory.java
package ch.bailu.aat.services.directory;

import java.io.IOException;
import java.util.Map;
import java.util.NavigableMap;
import java.util.TreeMap;

import ch.bailu.util_java.foc.Foc;

public class FilesInDirectory {

    final NavigableMap<String,Foc> files;


    public FilesInDirectory(Foc directory) throws IOException {
        files = getFileList(directory);

    }



    private static NavigableMap<String,Foc> getFileList(Foc directory) {
        final NavigableMap<String, Foc> files = new TreeMap<>();

        directory.foreachFile(new Foc.Execute() {
            @Override
            public void execute(Foc child) {
                files.put(child.getName(), child);
            }
        });

        return files;
    }


    public Foc findItem(String name) {
        return files.get(name);
    }


    public Foc pollItem(Foc file) {
        return files.remove(file.getName());
    }


    public Foc pollItem() {
        Map.Entry<String, Foc> e = files.pollLastEntry();

        if (e != null)
            return e.getValue();

        return null;
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/service/directory/FilesInDirectory.kt
package ch.bailu.aat_lib.service.directory

import ch.bailu.foc.Foc
import java.util.NavigableMap
import java.util.TreeMap

class FilesInDirectory(directory: Foc) {
    private val files: NavigableMap<String, Foc> = getFileList(directory)

    fun findItem(name: String): Foc? {
        return files[name]
    }

    fun pollItem(file: Foc): Foc? {
        return files.remove(file.name)
    }

    fun pollItem(): Foc? {
        return files.pollLastEntry()?.value
    }

    val size: Int
        get() = files.size

    companion object {
        private fun getFileList(directory: Foc): NavigableMap<String, Foc> {
            val files: NavigableMap<String, Foc> = TreeMap()
            directory.foreachFile { child: Foc -> files[child.name] = child }
            return files
        }
    }
}
